Change References to `state.actors()` to `ray.util.state.list_actors` or `ray.util.state.get_actor`

jcarlson212 opened the pull request:
## Why are these changes needed?

Changes references to `_private.state.actors()` to references to `ray.util.state.list_actors` or `ray.util.state.get_actor` depending on the plurality of the fetch. This helps the `_private` submodule better serve its purpose of containing abstractions only used for core ray internals. Note that `ray.runtime_context` still uses `_private.actors` as it's a core ray internal that needs to function regardless of whether or not the dashboard server is running.

## Related issue number

Fixes #53478 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

jcarlson212 committed:
Move `state.actors()` to `_common`

Signed-off-by: Jason <jcarlson212@gmail.com>

jcarlson212 commented on python/ray/_common/state.py:
```diff
@@ -0,0 +1,30 @@
+from typing import Optional
+
+import ray
+from ray._private.state import state
```
This is just a partial refactor. Scope of this PR is just this one function.

edoakes reviewed (changes requested):
I don't think we should move this one actually -- instead we should replace usage of it with the [state API](https://docs.ray.io/en/latest/ray-observability/reference/doc/ray.util.state.list_actors.html#ray.util.state.list_actors)

edoakes commented:
> @edoakes The [state API](https://docs.ray.io/en/latest/ray-observability/reference/doc/ray.util.state.list_actors.html#ray.util.state.list_actors) would require the dashboard to be running. Even though the dashboard is highly recommended, shouldn't we be agnostic of this functionality? `ray._private.state.actors()` already is agnostic of it despite not being intended for external usage, making it a decent option for being migrated to `_common` and referenced instead by other modules.

All of the usage changed in this PR is in test files, so I see no strong reason to avoid requiring the dashboard for these. It's a common pattern to use the state API for other libraries' tests as well.

jcarlson212 committed:
Use Public State API for Fetching Actors

Instead of fetching actors from the internal state API / migrating it to
the common module, simply reference the public state API that relies on the dashboard
server.

Signed-off-by: Jason <jcarlson212@gmail.com>

jcarlson212 commented:
> > @edoakes The [state API](https://docs.ray.io/en/latest/ray-observability/reference/doc/ray.util.state.list_actors.html#ray.util.state.list_actors) would require the dashboard to be running. Even though the dashboard is highly recommended, shouldn't we be agnostic of this functionality? `ray._private.state.actors()` already is agnostic of it despite not being intended for external usage, making it a decent option for being migrated to `_common` and referenced instead by other modules.
> 
> All of the usage changed in this PR is in test files, so I see no strong reason to avoid requiring the dashboard for these. It's a common pattern to use the state API for other libraries' tests as well.

There is a small amount of usage in the dataset (for split when hints are given) and runtime_context files as well. Will have the changes you requested posted soon though, just running some tests.

jcarlson212 committed:
Resolve Buildkite and Cursor Issues

Signed-off-by: Jason <jcarlson212@gmail.com>

jcarlson212 committed:
Fix Tests

Signed-off-by: Jason <jcarlson212@gmail.com>

jcarlson212 commented on python/ray/data/dataset.py:
```diff
@@ -2086,13 +2086,16 @@ def build_block_refs_by_node_id(
+                    + 100  # fetch current actors. Some staleness is assumed fine.
+                )
             }
+            return {actor: actors_state.get(actor._actor_id.hex()) for actor in actors}
```
The actor.actor_id and actor._actor_id.hex() expressions should match. The limit doesn't have to match for correctness as some freshness is implicitly tolerable for i.e. linearizability.

jcarlson212 committed:
Ensure `ray.util.state` Module is Available in Test Envs

Signed-off-by: Jason <jcarlson212@gmail.com>

jcarlson212 committed:
Add detail=True for NodeId and NumRestarts

Signed-off-by: Jason <jcarlson212@gmail.com>

jcarlson212 committed:
Override Ruff for Module Import

Signed-off-by: Jason <jcarlson212@gmail.com>

jcarlson212 committed:
Resolve Remaining Import State Module Errors on Remote Workers

Signed-off-by: Jason <jcarlson212@gmail.com>

edoakes commented on python/ray/runtime_context.py:
```diff
@@ -405,8 +405,10 @@ def was_current_actor_reconstructed(self):
         assert (
             not self.actor_id.is_nil()
         ), "This method should't be called inside Ray tasks."
-        actor_info = ray._private.state.actors(self.actor_id.hex())
-        return actor_info and actor_info["NumRestarts"] != 0
```
this is a core API that needs to function w/o the dashboard, so we should leave it as-is

note that it's ok for core files in `python/ray/*.py` to use private APIs

edoakes commented on python/ray/tests/conftest.py:
```diff
@@ -62,6 +62,14 @@ def pre_envs(monkeypatch):
     yield
 
 
+def pytest_configure(config):
```
state should be available in everything except the "minimal" build (which has a separate CI test)

edoakes replied:
I don't think we need this; there are many other similar soft dependencies for which we don't have such explicit enforcement

I think this will also fail in the "minimal" build, which explicitly does not support the state API

edoakes commented on python/ray/tests/test_asyncio.py:
```diff
@@ -239,8 +244,10 @@ async def loop_forever(self):
     # Using ray task so it works in Ray Client as well.
     @ray.remote
     def check_actor_gone_now():
+        import ray.util.state as STATE
```
ditto, no import alias please

jcarlson212 committed:
Improve Import Hygiene and Cut RuntimeContext DepOn Dashboard

Signed-off-by: Jason <jcarlson212@gmail.com>

jcarlson212 committed:
Fix Runtime Context State Submodule Loading Bug

Fixes the runtime context state submodule loading bug by importing
the state API module as STATE at the top of the file. Uses similar safety precaution in
the dataset.py file since these were the only two public API spots recently changed (not isolated to
just tests).

Signed-off-by: Jason <jcarlson212@gmail.com>

jcarlson212 committed:
Remove State Import Aliases

Signed-off-by: Jason <jcarlson212@gmail.com>

jcarlson212 committed:
Fix Shutdown Only Dashboard Initialization Bug

The dashboard wasn't being initialized before unit tests for shutdown_only fixtures
since the params get ignored, causing some unit tests to fail.

Signed-off-by: Jason <jcarlson212@gmail.com>

jcarlson212 committed:
Add Back _private Actors Timestamps Unit Test

Signed-off-by: Jason <jcarlson212@gmail.com>

jcarlson212 committed:
Add Ray Dashboard Init To Failing Actor Tests

Signed-off-by: Jason <jcarlson212@gmail.com>

jcarlson212 committed:
Add Dashboard to Fix More Tests

Signed-off-by: Jason <jcarlson212@gmail.com>

jcarlson212 commented:
`ray/tests/test_global_gc.py::test_local_gc_called_once_per_interval` seems to be failing from timing issues not directly related to these changes @edoakes

edoakes commented on python/ray/_private/state.py:
```diff
@@ -988,7 +988,6 @@ def actors(
     actor_state_name: Optional[str] = None,
 ):
     """Fetch actor info for one or more actor IDs (for debugging only).
-
```
add back whitespace

edoakes commented on python/ray/data/dataset.py:
```diff
@@ -2086,13 +2087,15 @@ def build_block_refs_by_node_id(
-                .get("Address", {})
-                .get("NodeID")
-                for actor in actors
+            actors_state = {
+                actor.actor_id: actor.node_id
+                for actor in list_actors(
+                    detail=True,
+                    limit=summarize_actors().get("cluster", {}).get("total_actors", 0)
+                    + 100,  # fetch current actors. Some staleness is assumed fine.
+                )
             }
+            return {actor: actors_state.get(actor._actor_id.hex()) for actor in actors}
```
@bveeramani can you verify this is OK to change? Looks like a pretty core change to data scheduling logic.

jcarlson212 replied:
Small comment for @bveeramani - even before these changes in theory the number of actors could change after the map was built, so this shouldn't change any consistency guarantees from what we had before per my understanding (except one property where we might return a number of actors that does not correspond to any discrete point in time in CPU cycles, which I don't think would be relied on... we'd instead be interpolating between CPU cycles).

bveeramani replied:
>  Looks like a pretty core change to data scheduling logic.

I think this logic is only used by the `Dataset.split()` API, which isn't a particularly important API  because people use `Dataset.streaming_split()` instead.

> can you verify this is OK to change?

This looks reasonable. 

Is `list_actors ` comparably fast to `ray._private.state.actors()`? Users expect dataset composition to be fast. If `list_actors` can be slow (many seconds), that might cause confusion/friction for users.

Other than that, some minor suggestions:

- It wasn't obvious to me what the keys and values for `actors_state` represent based on a name. Something like `actor_hex_to_node_id` might be clearer
- What happens if we remove the `+ 100`? I think it's a good practice to avoid magic numbers/arbitrary constants if possible.

edoakes commented on python/ray/tests/test_actor.py:
```diff
@@ -1685,7 +1685,7 @@ def test_get_actor_after_same_name_actor_dead(shutdown_only):
     NAMESPACE_NAME = "test_namespace"
 
-    ray.init(namespace=NAMESPACE_NAME)
+    ray.init(namespace=NAMESPACE_NAME, include_dashboard=True)
```
was `include_dashboard` required? I think it should be true by default

edoakes reviewed (commented):
Looks good. Only real concern is the change in `data/dataset.py`.

Given it is a change to the scheduler of Ray Data and the rest of the changes here are test-only, it would probably be wise to leave that change out of this PR and update it separately (with the Ray Data maintainers' blessing).

jcarlson212 committed:
Remove Redundant Imports, Dataset.py Ref Revert, Improve Hygeine

Signed-off-by: Jason <jcarlson212@gmail.com>

edoakes committed:
Merge branch 'master' into master

jcarlson212 commented on python/ray/serve/tests/test_logging.py:
```diff
@@ -805,7 +805,7 @@ def test_start_serve_with_logging_config(self, serve_and_ray_shutdown):
 
         # Check proxy log
-        nodes = state_api.list_nodes()
+        nodes = list_nodes()
```
I can raise a small separate PR for fixing the ray serve `test_metrics#test_actor_summary`. The test doesn't seem to have ran and the issue is isolated to only this test case (doesn't affect ray running in production).

jcarlson212 replied:
Related PR: https://github.com/ray-project/ray/pull/56948
